### PR TITLE
Notify Bugsnag if customer fails to be created

### PR DESCRIPTION
**Pull request by sauloperez**
#### What? Why?

**TL;DR**: we just want to know when #6038 bug happens.

This will hopefully let us see in Bugsnag when #6038 happens and then we'll have all the diagnostics data. The fact that the `order.distributor` is set as optional but necessary to create the customer already gives us some clue.

Note this changed its intent due to code review. While initially implemented raising an exception when the customer creation failed, it's more conservative now. It sends a notification to Bugsnag when the bug occurs. No other behavior change.

I also took the chance to make a few improvements to specs I had to deal with which are not directly related to the bug.

#### What should we test?

Green build.

#### Release notes
Send a notification to Bugsnag if customer fails to be created
Changelog Category: Technical changes

### CODE REVIEW

**Comment by sauloperez** [comment]
It's become evident we need to move the customer creation logic to the **context** where it makes sense, somewhere at the time we place the order. Then, there'll be no need for so many conditionals.

**Review by andrewpbrett — APPROVED** [review]
Why is `distributor` optional for an order in the first place? What's the scenario where an order doesn't have a distributor?

**Comment by sauloperez** [comment]
I was wondering the same and I think that's because of the sheer amount of
responsibilities the order model has. Perhaps the state machine and
specifically the cart state?

El ds., 17 d’abr. 2021, 20:31, Andy Brett ***@***.***> va
escriure:

File: app/models/spree/order.rb [hunk 1]
```diff
@@ -756,10 +756,14 @@ def ensure_customer
       return if associate_customer
 
       customer_name = bill_address.andand.full_name
-      self.customer = Customer.create(enterprise: distributor, email: email_for_customer,
-                                      user: user, name: customer_name,
-                                      bill_address: bill_address.andand.clone,
-                                      ship_address: ship_address.andand.clone)
+      self.customer = Customer.create!(
```
Comment (mkllnk) [inline app/models/spree/order.rb:759]: Good point! The create could fail. I'm a bit nervous about this failing the whole checkout instead of just reporting to Bugsnag. But it's probably worth it for the investigation.

**Review by mkllnk — APPROVED** [review]

**Comment by mkllnk** [comment]
> Why is distributor optional for an order in the first place? What's the scenario where an order doesn't have a distributor?

Historical reasons. The order model came from Spree which doesn't know anything about enterprises. We added enterprises on top.

File: spec/factories/order_factory.rb [hunk 2]
```diff
@@ -10,6 +10,7 @@
     bill_address
     completed_at { nil }
     email { user&.email || customer.email }
+    distributor factory: :distributor_enterprise
```
Comment (mkllnk) [inline spec/factories/order_factory.rb:13]: This is causing a lot of spec failures.

File: app/models/spree/order.rb (hunk 1 above)
Comment (sauloperez) [reply app/models/spree/order.rb:759]: yes, but if this was the reason, it's not happening a whole lot. At least, that we know of.

**Comment by sauloperez** [comment]
I found the root cause; when placing an order from `/admin/orders/new` the order is persisted when adding items to it, just before moving to the customer details page `admin/orders/R147857134/customer`. There's no user or `order.email` at that point (see [Order#email_for_customer](https://github.com/openfoodfoundation/openfoodnetwork/blob/9bddf5b00f398f83aefa14da6b1c3d2d8539f312/app/models/spree/order.rb#L746-L764))

File: app/controllers/spree/admin/orders_controller.rb [hunk 3]
```diff
@@ -26,6 +26,7 @@ def new
       def edit
         @order.shipments.map(&:refresh_rates)
 
+        @order.customer_creation = false
```
Comment (sauloperez) [inline app/controllers/spree/admin/orders_controller.rb:29]: this gives us a clue. Backoffice orders may be the way to reproduce this.

**Comment by mkllnk** [comment]
@sauloperez I'm a bit confused about the purpose of this pull request now. I think that you did some really good analysis and you found a way to create an order without customer record. But then you also added a way to skip the creation of a customer record which doesn't help to solve the issue, or does it?

Should this be in Code Review or In Dev?

**Comment by sauloperez** [comment]
Because the feature spec covering /admin/orders/new failed to create orders 🤷 . Before this change they where likely failing silently. I'll check it again to see if I'm right.

File: spec/serializers/api/admin/order_serializer_spec.rb [hunk 4]
```diff
@@ -67,4 +68,24 @@
       end
     end
   end
+
+  describe "#completed_at" do
+    let(:order) { build(:order, state: 'complete', completed_at: DateTime.parse("2021-04-02")) }
```
Comment (Matt-Yorkley) [inline spec/serializers/api/admin/order_serializer_spec.rb:73]: This is so much nicer than the deleted section in the controller spec :heart:

File: app/models/spree/order.rb (hunk 1 above)
Comment (Matt-Yorkley) [reply app/models/spree/order.rb:761]: I'm a bit nervous as well...

The "soft" option here would be to drop the bang! and add:
```ruby
Bugsnag.notify("Customer not created with order") unless customer.persisted?
```
that way if we encounter this we'll get an alert with a backtrace instead of a potentially checkout-blocking error, right?

File: app/models/spree/order.rb [hunk 5]
```diff
@@ -733,6 +733,7 @@ def skip_payment_for_subscription?
     end
 
     def require_customer?
+      return false if customer_creation == false
       return false if new_record? || state == 'cart'
```
Comment (Matt-Yorkley) [inline app/models/spree/order.rb:737]: Isn't the order in cart state when it's at the "customer creation" step? Or is it later? Address state?

Comment (sauloperez) [reply app/models/spree/order.rb:737]: The one just after, address, exactly. All this machinery we have is to implement that "conditional"

File: app/models/spree/order.rb (hunk 1 above)
Comment (sauloperez) [reply app/models/spree/order.rb:761]: fair enough. It's not that the current situation is much better than a checkout-blocking error. At least, as is, the hub doesn't miss a "conversion", that's true.

File: app/models/spree/order.rb (hunk 5 above)
Comment (Matt-Yorkley) [reply app/models/spree/order.rb:737]: I'm just wondering if we should restrict it a bit, like we're now unconditionally setting `@order.customer_creation = false` in the edit action before advancing the order steps with `OrderWorkflow`, right? And that'll mean that `#require_customer?` will always return `false`, regardless of the order state? Or no?

Would this be preferable in the edit action? :point_down: 
```ruby
@order.customer_creation = false if @order.address?
```
Sorry for all the questions! I'm just trying to make sense of it in my head :thinking:

File: app/models/spree/order.rb (hunk 1 above)
Comment (sauloperez) [reply app/models/spree/order.rb]: @Matt-Yorkley I just changed the approach as you mentioned. Let me know what you think. If we like it might squash some commits before moving on.

File: app/models/spree/order.rb (hunk 5 above)
Comment (sauloperez) [reply app/models/spree/order.rb:737]: This is not needed now with the change in approach, because nothing blows up if the customer fails to be created, but you do well to ask.

At the point where I just added an item in the cart :point_down: , a request is sent to save it. First, `#requier_customer?` is checked and it does return true. That's because `new_record? ==> false` and `state == "address"`. This added `customer_creation` flag avoided this.

**Review by Matt-Yorkley — APPROVED** [review]
:+1:

File: app/models/spree/order.rb (hunk 5 above)
Comment (Matt-Yorkley) [reply app/models/spree/order.rb]: Wait, now we're not doing this at all? I'm confused :see_no_evil: I need to drink more coffee and re-read it :sweat_smile:

Comment (sauloperez) [reply app/models/spree/order.rb]: dude! you made me change the approach and simply notify Bugsnag :sweat_smile: :joy:

Comment (Matt-Yorkley) [reply app/models/spree/order.rb]: Ah, I assumed we would do the "soft" Bugnsag thing, but keep the `@order.customer_creation = false` idea, so the order can progress to the next step? :joy:

Comment (sauloperez) [reply app/models/spree/order.rb]: Now I see. This is the misunderstanding. It does progress to the next step right now, but leaves the order without an associated customer up until the next page, where the user fills up the customer form.

File: app/models/spree/order.rb [hunk 6]
```diff
@@ -756,11 +756,19 @@ def associate_customer
...
+        enterprise: distributor,
+        email: email_for_customer,
+        user: user,
+        name: bill_address.andand.full_name,
+        bill_address: bill_address.andand.clone,
+        ship_address: ship_address.andand.clone
+      )
+      customer.save
+
+      unless customer.valid?
```
Comment (andrewpbrett) [inline app/models/spree/order.rb:769]: Very minor point but I'd write this as a one-liner with `if customer.errors.any?` since we expect the customer to be valid.

**Review by andrewpbrett — APPROVED** [review]
This is now pretty simple, looks good to me 👍

File: app/models/spree/order.rb (hunk 6 above)
Comment (sauloperez) [reply app/models/spree/order.rb]: I pushed-forced a new commit making this conditional "positive" without having to couple to `errors`

File: app/models/spree/order.rb [hunk 7]
```diff
@@ -756,11 +756,19 @@ def associate_customer
...
+        enterprise: distributor,
+        email: email_for_customer,
+        user: user,
+        name: bill_address.andand.full_name,
+        bill_address: bill_address.andand.clone,
+        ship_address: ship_address.andand.clone
+      )
+      customer.save
+
+      if customer.invalid?
```
Comment (Matt-Yorkley) [inline app/models/spree/order.rb:769]: Why not check `#persisted?`. I think it avoids a query (and it's more explicitly what we're looking to check, righ?)...

```ruby
irb(main):006:0> customer.persisted?
=> false
irb(main):007:0> customer.valid?
  Spree::Customer Exists (1.6ms)  SELECT  1 AS one FROM "spree_customers" WHERE "spree_customers"."email" = $1 LIMIT $2  [["email", "tim@test.com"], ["LIMIT", 1]]
=> false
irb(main):008:0> customer.invalid?
  Spree::Customer Exists (1.3ms)  SELECT  1 AS one FROM "spree_customers" WHERE "spree_customers"."email" = $1 LIMIT $2  [["email", "tim@test.com"], ["LIMIT", 1]]
=> true
```

Comment (sauloperez) [reply app/models/spree/order.rb:769]: but `invalid?` triggers the model validations (AFAIK persisted? does not) and it's more intention revealing. Besides, we aiming to get this code out to assess the situation. It's but no means a solution.

Comment (sauloperez) [reply app/models/spree/order.rb:769]: I find more intention revealing to check for valid?/validation?. What we care for is whether or not the customer passed the validations and if not, we'll list which ones failed.

Besides, this is by no means a solution. We're just trying to get some more insights but I'll see saves unnecessary queries anway.

**Review by Matt-Yorkley — APPROVED** [review]

**Comment by sauloperez** [comment]
My mistake. This wasn't meant to be manually tested but let's see if you can find anything @filipefurtad0

**Comment by filipefurtad0** [comment]
No worries @sauloperez :+1: 
I started off by doing a quick test, but I found it could be useful to dig in a bit here.

- shopfront orders

Seem to work fine :+1: Could not trigger any errors here. (there are some limitations on staging-ES, so I might re-test this on another server)

- Backoffice orders

While filling the details, I noticed that Bugsnag was triggered, between the `Order Details` section and the `Customer Details` or, to put it more precisely, after adding line items to the order:

image

Here is the error:
https://app.bugsnag.com/katuma/katuma/errors/6086d970aa49cf0007233683?event_id=6086d9700078226222d70000&i=sk&m=nw

I'd say that at this stage it is perfectly fine to not to have a customer address, so we should be aware that this error is triggered in a regular situation, before the user even got to fill in customer details.

However, there is something weird going on about the order states, after this PR. Orders with non-captured payments have `shippment state `and `payment state` = NULL, see below (red, this PR; green, expected behavior)

image

Orders with `payment_state=NULL` were previously known to break the payment reports (in the context of the previous Rails upgrade, see #7021) - and it does -> visiting /admin/reports/payments and selecting Payment Totals -> Snail

https://app.bugsnag.com/katuma/katuma/errors/6087c477d3c10100070539c4?event_id=6087c477007815009fbd0000&i=sk&m=nw

I'm not sure we should merge this, without assuring it won't break reports -> indeed a pain point in testing and specs :face_with_head_bandage:

**Comment by Matt-Yorkley** [comment]
Ah... I think that recurrence of issue #7021 isn't introduce by this PR, and I think it might be in the release. We need to take a look at that before deploying.

**Comment by filipefurtad0** [comment]
Indeed,
the strange behavior in order states is not introduced by this PR, it is in master. The comparison pic above is actually from Rails 5.2, which matched the expected behavior.

Great catch @Matt-Yorkley! I'll open an issue, and comment on the release issue

Also, this probably means we're good to go on this PR :+1: 

Thanks for the heads up!

**Comment by sauloperez** [comment]
Forever grateful @Matt-Yorkley . You saved me there 😂. And @filipefurtad0, thanks a lot for the thoroughness 🏅
